zerver/tests/test_markdown.py

```diff
@@ -395,6 +396,144 @@ def test_silent_mention_user_groups(self) -> None:
+        )
+
+    def test_fetch_group_membership_for_bulk_mention_groups(self) -> None:
+        # Same as test_fetch_group_membership_when_mention_group_permission_changes
+        # but for bulk mentioning groups in a single message.
+        # This test ensures we do the permission check for bulk groups mentions correctly
+        # and in a constant time.
+        realm = get_realm("zulip")
+
+        hamlet = self.example_user("hamlet")
+        cordelia = self.example_user("cordelia")
+        hamlet_members_ids = {hamlet.id, cordelia.id}
+
+        iago = self.example_user("iago")
+        aaron = self.example_user("aaron")
+        zoe = self.example_user("ZOE")
+        othello = self.example_user("othello")
+
+        can_mention_group = check_add_user_group(
+            realm, "can_mention_group", [aaron], acting_user=iago
+        )
+        moderators_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MODERATORS, realm=realm, is_system_group=True
+        )
+        sub_group = check_add_user_group(realm, "sub_group_1", [zoe], acting_user=iago)
+        add_subgroups_to_user_group(can_mention_group, [sub_group], acting_user=iago)
+        hamlet_group = NamedUserGroup.objects.get(realm=realm, name="hamletcharacters")
+
+        content = "@*hamletcharacters*, "
+        mentioned_groups_ids = []
+        for i in range(6):
+            group_name = f"group_{i}"
+            group = check_add_user_group(realm, group_name, [othello], acting_user=iago)
+            mentioned_groups_ids.append(group.id)
+            content += f"@*{group_name}*, "
+
+            # We allow some groups to be mentioned by the sender "zoe" through can_mention_group,
+            # but prevent other groups through moderators_group.
+            # This way we can test that our algorithm correctly and efficiently does the permission
+            # checks against different mention permissions for different groups.
+            if i % 2 == 0:
+                do_change_user_group_permission_setting(
+                    group,
+                    "can_mention_group",
+                    can_mention_group,
+                    acting_user=None,
+                )
+            else:
+                do_change_user_group_permission_setting(
+                    group,
+                    "can_mention_group",
+                    moderators_group,
+                    acting_user=None,
+                )
+
+        mention_backend = MentionBackend(realm.id)
+
+        # Ensure we are doing the mention permission checks against multiple
+        # groups in a constant time.
+        # This is one extra query than test_bulk_user_group_mentions,
+        # that extra query is only triggered when any mentioned group
+        # needs permission check (e.g. not SystemGroups.EVERYONE).
+        with self.assert_database_query_count(3):
+            mention_data = MentionData(mention_backend, content, message_sender=zoe)
```

Comment:
We should also have one test for checking query count when every mentioned group has can_mention_group as EVERYONE